### DISCUSSION TITLE
[@types/node] Node crypto.verify does not accept a JsonWebKey as key

### DISCUSSION THREAD

**Discussion by KantiKuijk** [DefinitelyTyped/DefinitelyTyped#64387] [Issues with a @types package]
A few months ago this issue got fixed regarding "jwk" being accepted as a valid `KeyFormat` in the crypto node module: https://github.com/DefinitelyTyped/DefinitelyTyped/discussions/63491.
The crypto node module can also create private/public keys given a json web key (jwk), named as the `JsonWebKey` type in @types/node with the `crypto.createPrivateKey` and `crypto.createPublicKey` functions respectively.

However, when trying to use the `crypto.verify` function and giving it an object from which it creates a public key with the aforementioned functions, typescript only allows `string` and `Buffer` as the key types, but not a `JsonWebKey`.

Thus although this typescript is valid with `jwk` being of type `JsonWebKey` : 
```javascript
const publicKey = crypto.createPublicKey({format: "jwk", key: jwk})
const verified = verify(someAlgorithm, someData, publicKey, someSignature)
```
this typescript is not valid, saying that `{typeof jwk} is not assignable to type 'string | Buffer'`:
```javascript
const verified = verify(
    someAlgorithm, someData,
    {format: "jwk", key: jwk},
    someSignature
)
```
Of course you could just use the first example, but then you can't add the properties `dsaEncoding`, `padding` or `saltLength` as described in the crypto docs: https://nodejs.org/api/crypto.html#cryptoverifyalgorithm-data-key-signature-callback.

I'm not well versed enough in all of this but I believe at least part of the problem lies in that `crypto.createPublicKey` typing accepts `JsonWebKeyInput` as type (among others). But `crypto.verify` only accepts `KeyLike | VerifyKeyObjectInput | VerifyPublicKeyInput` as types for the `key` parameter. So I believe making a `interface VerifyJsonWebKeyInput extends JsonWebKeyInput, SigningOptions {}` and adding it to the `key` parameter of `crypto.verify` to be `KeyLike | VerifyKeyObjectInput | VerifyPublicKeyInput | VerifyJsonWebKeyInput` might fix it.
But again I am not knowledgeable enough to be sure this will definitely fix it, or to be sure this is the correct way of fixing the issue (without breaking other stuff).

[a comment by an automated account is not shown]

**Reply by msalman360** [reply]
https://linksforeveryone.com/json-formatter/

**Comment by KantiKuijk** [comment]
Proposing a fix in https://github.com/DefinitelyTyped/DefinitelyTyped/pull/64467

**Reply by msalman360** [reply]
https://linksforeveryone.com/json-formatter/
